lang/java/avro/src/main/java/org/apache/avro/SchemaNormalization.java

```diff
@@ -147,14 +175,55 @@ private static Appendable build(Map<String,String> env, Schema s,
         for (Schema.Field f: s.getFields()) {
           if (! firstTime) o.append(','); else firstTime = false;
           o.append("{\"name\":\"").append(f.name()).append("\"");
-          build(env, f.schema(), o.append(",\"type\":")).append("}");
+          build(env, f.schema(), o.append(",\"type\":"), ps, aps);
+          if (!ps) writeFieldProps(o, f, aps); // if standard canonical form then add reserved properties
+          o.append("}");
         }
         o.append("]");
       }
+      if(!ps) { writeComplexProps(o, s); writeProps(o, s.getObjectProps(), aps); }// adding the reserved property if not parser canonical schema
       return o.append("}");
     }
   }
 
+  private static Appendable writeLogicalType(Schema s, LogicalType lt, Appendable o, LinkedHashSet<String> aps) throws IOException {
+    o.append("{\"type\":\"").append(s.getType().getName()).append("\"");
+    o.append("\"").append(LogicalType.LOGICAL_TYPE_PROP).append("\":\"").append(lt.getName()).append("\"");
+    if (lt.getName() == "decimal") {
+      LogicalTypes.Decimal dlt = (LogicalTypes.Decimal) lt;
+      o.append(",\"precision\":").append(Integer.toString(dlt.getPrecision()));
+      if(dlt.getScale() != 0) o.append(",\"scale\":").append(Integer.toString(dlt.getScale()));
+    }
+    // adding the reserved property
+    writeProps(o, s.getObjectProps(), aps);
+    return o.append("}");
+  }
+
+  private static Appendable writeProps(Appendable o, Map<String, Object> schemaProps, LinkedHashSet<String> aps) throws IOException {
+    for (String propKey : aps) {
+      if (schemaProps.containsKey(propKey)) {
+        String propValue = JacksonUtils.toJsonNode(schemaProps.get(propKey)).toString();
+        o.append(",\"").append(propKey).append("\":").append(propValue);
+      }
+    }
+    return o;
+  }
+
+  private static Appendable writeComplexProps(Appendable o, Schema s) throws IOException {
+    if (s.getDoc() !=null && !s.getDoc().isEmpty()) o.append(",\"doc\":\"").append(s.getDoc()).append("\"");
+    if (s.getAliases() !=null && !s.getAliases().isEmpty()) o.append(",\"aliases\":").append(JacksonUtils.toJsonNode(s.getAliases()).toString());
```

Comment:
Should the values for aliases be sorted when outputting the canonical? I don't believe that the order is important in the schema. Without sorting, two versions of the schema that just differ in the order of the aliases would be considered different canonical schemas.

Reply:
@tjwp Thanks for review comments. We have the different canonical schema for `aliases` property value unordered. Do you think, any other property have similar cases ?

Reply:
@cutting @tjwp  as per AVRO standard we are maintain `aliases` as `LinkedHashSet` it provides insertion order  Set. Can we keep it as it is, if user defined order of aliases will maintain in the canonical form of schema. If you think, any where we are maintain similar cases, let me know. Finally, we need to have same canonical schema.

Reply:
`aliases` is the only property where I think the string values need to be sorted for the canonical form. 

Reading back over the spec I also noticed:

```
A type alias may be specified either as a fully namespace-qualified, or relative to the namespace of the name it is an alias for. For example, if a type named "a.b" has aliases of "c" and "x.y", then the fully qualified names of its aliases are "a.c" and "x.y".
```

I think the implication of this is that `aliases` for types should also be normalized in the canonical form. That normalization would be to either always fully qualify the name, or always strip the namespace if it matches the namespace of the name that is aliased. Of the two options I'd favor always fully qualifying type aliases as that's consistent with the handling of names in the parsing canonical form.

An example:

This type:
```
{
  "name": "B",
  "namespace": "A",
  "type": "record",
  "fields": [
    { "name": "X", "type": "int", aliases: ["Y", "Z"] }
  ],
  "aliases": ["C.B", "D"]
}
```

And this type:

```
{
  "name": "B",
  "namespace": "A",
  "type": "record",
  "fields": [
    { "name": "X", "type": "int", aliases: ["Z", "Y"] }
  ],
  "aliases": ["C.B", "A.D"]
}
```

Should have the same canonical form (based on what I've proposed -- excluding whitespace which I've kept for readability):

```
{
  "name": "A.B",
  "type": "record",
  "fields": [
    { "name": "X", "type": "int", aliases: ["Y", "Z"] }
  ],
  "aliases": ["A.D", "C.B"]
}
```

Reply:
@tjwp  as per avro  `NamedSchema` and `Field` has `aliases` property. How this behaviour should order for nested RECORD type ? `namespace` of inner record type field is optional (may or may not be provided), so lot of scenarios we have to handle this ordering.
```
{ 
  "type": "record",
  "name": "person",
  "namespace": "com.test",
  "fields": [
    {
      "name": "firstname",
      "type": "string",
      "aliases": ["name", "f_n"]
    },
    {
      "name": "lastname",
      "type": "string",
      "aliases": ["l_n"]
    },
    {
      "name": "address",
      "type": {
		  "type": "record",
		  "name": "Address",
		  "fields": [
		    {
		      "name": "streetaddress",
		      "type": "string"
		    },
		    {
		      "name": "city",
		      "type": "string"
		    }
		  ],
		  "aliases": ["Location", "Add"]
		}
    }
  ],
  "aliases": ["com.mnp.person", "com.abc.person", "employee"]
}

Reply:
Namespacing only applies to the names (and aliases) of records, not fields. So the aliases of fields just need to be ordered.

If a namespace is specified for a record name, or alias, then it is a full namespace so we don't have to worry about that type of nesting.

With nesting, the "namespace is taken from the most tightly enclosing schema". A simple way of looking at this is that whatever namespace applies to the record name should be applied to aliases (that are not already fully qualified) for that record.

For your example, this looks like (again ignoring whitespace):

```
{ 
  "name": “com.test.person",
  "type": "record”,
  "fields": [
    {
      "name": "firstname",
      "type": "string",
      "aliases": [“f_n”, "name"]
    },
    {
      "name": "lastname",
      "type": "string",
      "aliases": ["l_n"]
    },
    {
      "name": "address",
      "type": {
                  "name": “com.test.Address",
                  "type": "record”,
                  "fields": [
                    {
                      "name": "streetaddress",
                      "type": "string"
                    },
                    {
                      "name": "city",
                      "type": "string"
                    }
                  ],
                  "aliases": [“com.test.Add”, “com.test.Location"]
                }
    }
  ],
  "aliases": ["com.abc.person”, "com.mnp.person", “com.test.employee"]
}
```

Reply:
@tjwp code changes for aliases ordering also done. Can you review again ?

Reply:
@tjwp you can see the test cases sample here. Let me know, i will work on spec.
```
// 011
<<INPUT
  {"name":"PigValue","type":"record","fields":[{"name":"value", "type":"string", "format": { "key": "value" }, "aliases": ["B", "A"]}]}
INPUT
<<canonical {"name":"PigValue","type":"record","fields":[{"name":"value","type":"string","order":"ASCENDING","aliases":["A","B"],"format":{"key":"value"}}]}

// 012
<<INPUT
  {"namespace":"com.avro","name":"PigValue","type":"record","aliases":["sample", "com.sample.PVal"],"fields":[{"name":"value", "type":"string", "format": { "key": "value" }}]}
INPUT
<<canonical {"name":"com.avro.PigValue","type":"record","fields":[{"name":"value","type":"string","order":"ASCENDING","format":{"key":"value"}}],"aliases":["com.avro.sample","com.sample.PVal"]}

// 013
<<INPUT
  {"aliases":["sample", "com.sample.PVal"],"name":"PigValue","type":"record","fields":[{"name":"value", "aliases": ["B", "A"], "type":"string", "format": { "key": "value" }}]}
INPUT
<<canonical {"name":"PigValue","type":"record","fields":[{"name":"value","type":"string","order":"ASCENDING","aliases":["A","B"],"format":{"key":"value"}}],"aliases":["com.sample.PVal","sample"]}

// 014
<<INPUT
  {"namespace":"com.avro","name":"PigValue","type":"record","aliases":["sample", "com.sample.PVal"],"fields":[{"name":"value", "type":"string", "format": { "key": "value" }},{"name":"nested","type":{"type":"record","name":"Nested","fields":[ ],"aliases":["L","A"]}}]}
INPUT
<<canonical {"name":"com.avro.PigValue","type":"record","fields":[{"name":"value","type":"string","order":"ASCENDING","format":{"key":"value"}},{"name":"nested","type":{"name":"com.avro.Nested","type":"record","fields":[],"aliases":["com.avro.A","com.avro.L"]},"order":"ASCENDING"}],"aliases":["com.avro.sample","com.sample.PVal"]}